File: Trapping Rainwater/Leetcode_42.py
# top 1%
class Solution:
    def trap(self, height: List[int]) -> int:
        left, right = 0, len(height)-1
        maxLeft, maxRight = 0, 0
        ans = 0
        while left <= right:
            if height[left] <= height[right]:
                if height[left] >= maxLeft:
                    maxLeft = height[left]
                else:
                    ans += maxLeft - height[left]
                left += 1
            else: # height[right] > height[left]
                if height[right] >= maxRight:
                    maxRight = height[right]
                else:
                    ans += maxRight - height[right]
                right -= 1
        return ans

File: Trapping Rainwater/Leetcode_42.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int ans = 0;
        int l = 0, r = height.size() - 1;
        int lMax = 0, rMax = 0;
        while (l<r){
            lMax = max(lMax, height[l]);
            rMax = max(rMax, height[r]);
            if (lMax < rMax) ans += max(0, lMax - height[l++]);
            else ans += max(0, rMax - height[r--]);
        }
        return ans;
    }
};
